core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,263 @@
+  }
+
+  public static Builder buildFrom(SnapshotOperations base) {
+    return new Builder(base);
+  }
+
+  public static Builder buildFromEmpty() {
+    return new Builder();
+  }
+
+  public static SnapshotOperations empty() {
+    return SnapshotOperations.buildFromEmpty().build();
+  }
+
+  static class Builder {
+    private List<Snapshot> snapshots;
+    private SerializableSupplier<List<Snapshot>> snapshotsSupplier;
+    private final Map<Long, Snapshot> snapshotsById;
+    private Map<String, SnapshotRef> refs;
+    private SerializableSupplier<Map<String, SnapshotRef>> refsSupplier;
+
+    Builder() {
+      this.snapshots = Lists.newArrayList();
+      this.refs = Maps.newHashMap();
+
+      this.snapshotsById = Maps.newHashMap();
+    }
+
+    Builder(SnapshotOperations base) {
+      this.snapshots = Lists.newArrayList(base.snapshots);
+      this.snapshotsSupplier = base.snapshotsSupplier;
+      this.refs = Maps.newHashMap(base.refs);
+      this.refsSupplier = base.refsSupplier;
+
+      this.snapshotsById = indexSnapshotsById(snapshots);
+    }
+
+    Builder snapshots(List<Snapshot> snapshots1) {
+      this.snapshots = snapshots1;
+      return this;
+    }
+
+    Builder snapshotsSupplier(SerializableSupplier<List<Snapshot>> snapshotsSupplier1) {
+      this.snapshotsSupplier = snapshotsSupplier1;
+      return this;
+    }
+
+    Snapshot snapshot(Long id) {
+      return snapshotsById.get(id);
+    }
+
+    Builder add(Snapshot snapshot) {
+      ValidationException.check(
+          !snapshotsById.containsKey(snapshot.snapshotId()),
+          "Snapshot already exists for id: %s",
+          snapshot.snapshotId());
+
+      snapshots.add(snapshot);
+      snapshotsById.put(snapshot.snapshotId(), snapshot);
+
+      return this;
+    }
+
+    List<Snapshot> remove(Collection<Long> idsToRemove) {
+      List<Snapshot> retainedSnapshots =
+          Lists.newArrayListWithExpectedSize(snapshots.size() - idsToRemove.size());
+      List<Snapshot> removedSnapshots = Lists.newArrayListWithExpectedSize(idsToRemove.size());
+
+      for (Snapshot snapshot : snapshots) {
+        long snapshotId = snapshot.snapshotId();
+        if (idsToRemove.contains(snapshotId)) {
+          Snapshot removed = snapshotsById.remove(snapshotId);
+          removedSnapshots.add(removed);
+        } else {
+          retainedSnapshots.add(snapshot);
+        }
+      }
+
+      this.snapshots = retainedSnapshots;
+
+      // remove any refs that are no longer valid
+      Set<String> danglingRefs = Sets.newHashSet();
+      for (Map.Entry<String, SnapshotRef> refEntry : refs.entrySet()) {
+        if (!snapshotsById.containsKey(refEntry.getValue().snapshotId())) {
+          danglingRefs.add(refEntry.getKey());
+        }
+      }
+
+      danglingRefs.forEach(this::removeRef);
+
+      return removedSnapshots;
+    }
+
+    Builder refs(Map<String, SnapshotRef> refs) {
```

Comment:
Since this builder has methods that return data, I think it isn't very clear to have `ref` that returns a `SnapshotRef` and `refs` that sets all of the refs. Can you rename this to `setRefs`?